sensor/FamilyProtectPlugin.js

```diff
@@ -0,0 +1,193 @@
+                if (macAddress) {
+                    if (policy == true) {
+                        this.enabledMacAddresses[macAddress] = 1;
+                    } else {
+                        delete this.enabledMacAddresses[macAddress];
+                    }
+                    return this.applyDeviceFamilyProtect(macAddress);
+                }
+            }
+        } catch (err) {
+            log.error("Got error when applying family protect policy", err);
+        }
+    }
+
+    async applyFamilyProtect() {
+        await this.applySystemFamilyProtect();
+        for (const macAddress in this.enabledMacAddresses) {
+            await this.applyDeviceFamilyProtect(macAddress);
+        }
+    }
+
+    async applySystemFamilyProtect() {
+        this.familyDnsAddr((err, dnsaddrs) => {
+            if (this.systemSwitch && this.adminSystemSwitch) {
+                return this.systemStart(dnsaddrs);
+            } else {
+                return this.systemStop(dnsaddrs);
+            }
+        });
+    }
+
+    async applyDeviceFamilyProtect(macAddress) {
+        this.familyDnsAddr((err, dnsaddrs) => {
+            try {
+                if (this.enabledMacAddresses[macAddress] && this.adminSystemSwitch) {
+                    return this.perDeviceStart(macAddress, dnsaddrs)
+                } else {
+                    return this.perDeviceStop(macAddress, dnsaddrs);
+                }
+            } catch (err) {
+                log.error(`Failed to apply family protect on device ${macAddress}, err: ${err}`);
+            }
+        });
+    }
+
+    async systemStart(dnsaddrs) {
+        dnsmasq.setDefaultNameServers("family", dnsaddrs);
+        dnsmasq.updateResolvConf();
+    }
+
+    async systemStop(dnsaddrs) {
+        dnsmasq.unsetDefaultNameServers("family"); // reset dns name servers to null no matter whether iptables dns change is failed or successful
+        dnsmasq.updateResolvConf();
+    }
+
+    async perDeviceStart(macAddress, dnsaddrs) {
+        const configFile = `${dnsmasqConfigFolder}/familyProtect_${macAddress}.conf`;
+        const dnsmasqentry = `server=${dnsaddrs[0]}%${macAddress.toUpperCase()}\n`;
+        await fs.writeFile(configFile, dnsmasqentry);
+        dnsmasq.restartDnsmasq();
+    }
+
+    async perDeviceStop(macAddress, dnsaddrs) {
+        const configFile = `${dnsmasqConfigFolder}/familyProtect_${macAddress}.conf`;
+        await fs.unlink(configFile, err => {
+            if (err) {
+                if (err.code === 'ENOENT') {
+                    log.info(`Dnsmasq: No ${configFile}, skip remove`);
+                } else {
+                    log.warn(`Dnsmasq: Error when remove ${configFile}`, err);
+                }
+            }
+        })
+        dnsmasq.restartDnsmasq();
+    }
+
+    // global on/off
+    async globalOn() {
+        this.adminSystemSwitch = true;
+        await this.applySystemFamilyProtect();
```

Comment:
I think this.applyFamilyProtect should be called here which will apply both system and device level changes. The feature switch should switch on/off system level and device level simultaneously.